File: backend/src/branch.cpp
#include <glibmm.h>

#include "branch.hpp"

using namespace std;
using namespace Ats;

unique_ptr<Branch>
Branch::create(std::string type, uint stream, uint chan, uint pid) {

    if (type == "video") return unique_ptr<Branch>((Branch*) new Video_branch(stream, chan, pid));
    else if (type == "audio") return unique_ptr<Branch>((Branch*) new Audio_branch(stream, chan, pid));
    else return unique_ptr<Branch>(nullptr);
}

Branch::Branch() {
    _bin    = Gst::Bin::create();
    auto queue   = Gst::ElementFactory::create_element("queue");
    _decoder = Gst::ElementFactory::create_element("decodebin");

    queue->set_property("max-size-buffers", 20000);
    queue->set_property("max-size-bytes", 12000000);

    _bin->add(queue)->add(_decoder);
    queue->link(_decoder);
}

void
Branch::connect_src ( const Glib::RefPtr<Gst::Pad>& p ) {
    auto sink_pad  = _bin->get_static_pad("sink");
    p->link(sink_pad);
    _bin->sync_state_with_parent();
}

Video_branch::Video_branch(uint stream, uint chan, uint pid) {

    _decoder->signal_pad_added().
	connect([this,stream,chan,pid](const Glib::RefPtr<Gst::Pad>& pad)
		{
		    auto pcaps = pad->get_current_caps()->get_structure(0).get_name();
		    vector<Glib::ustring> caps_toks = Glib::Regex::split_simple("/", pcaps);
		    auto& type    = caps_toks[0];

		    if (type != "video") return;

		    auto deint  = Gst::ElementFactory::create_element("deinterlace");
		    auto _analyser = Gst::ElementFactory::create_element("videoanalysis");
		    
		    auto sink_pad = deint->get_static_pad("sink");
		    auto src_pad  = _analyser->get_static_pad("src");

		    _bin->add(deint)->add(_analyser);
		    deint->link(_analyser);

		    pad->link(sink_pad);

		    auto src_ghost = Gst::GhostPad::create(src_pad, "src");
		    src_ghost->set_active();
		    _bin->add_pad(src_ghost);

		    auto p = shared_ptr<Pad>(new Pad(stream,
						     chan, pid,
						     "video",
						     (Glib::RefPtr<Gst::Pad>) src_ghost));

		    _pads.push_back(p);
		    _pad_added.emit(p);
		});
    
}

Audio_branch::Audio_branch(uint stream, uint chan, uint pid) {

}


/* 
   TODO: MOVE TO WINDOW 
   void
   Graph::build_subbranch(RefPtr<Gst::Bin> bin, uint stream, uint channel, uint pid, const RefPtr<Gst::Pad>& p) {
    RefPtr<Gst::Pad> src_pad;

    auto set_video = [this,&p,stream,channel,pid](){
	Meta_pid::Video_pid v;

	auto vi = gst_video_info_new();
	auto pcaps = p->get_current_caps();
	if (! gst_video_info_from_caps(vi, pcaps->gobj())) {
	    gst_video_info_free(vi);
	    return;
	}
			
	v.codec = "h264"; // FIXME not only h264 supported
	v.width = vi->width;
	v.height = vi->height;
	v.aspect_ratio = {vi->par_n,vi->par_d};
	v.frame_rate = (float)vi->fps_n/vi->fps_d;

	gst_video_info_free(vi);

	Meta_pid::Pid_type rval = v;

	set_pid.emit(stream,channel,pid,rval);
    };

    auto pcaps = p->get_current_caps()->get_structure(0).get_name();
    vector<Glib::ustring> caps_toks = Glib::Regex::split_simple("/", pcaps);
    auto& type    = caps_toks[0];	    

    if (type == "video") {
	// Video callback
	p->connect_property_changed("caps", set_video);
	set_video();
		
	auto _deint  = Gst::ElementFactory::create_element("deinterlace");
	auto anal    = Gst::ElementFactory::create_element("videoanalysis");
	auto _scale  = Gst::ElementFactory::create_element("videoscale");
	auto _caps   = Gst::ElementFactory::create_element("capsfilter");

	_caps->set_property("caps", Gst::Caps::create_from_string("video/x-raw,pixel-aspect-ratio=1/1"));

	
//	n.type = type;
//	n.analysis = anal;
	
			
	auto _sink = _deint->get_static_pad("sink");
	src_pad = _caps->get_static_pad("src");
		
	bin->add(_deint)->add(anal)->add(_scale)->add(_caps);
	_deint->link(anal)->link(_scale)->link(_caps);
		
	_deint->sync_state_with_parent();
	anal->sync_state_with_parent();
	_scale->sync_state_with_parent();
	_caps->sync_state_with_parent();
		
	p->link(_sink);
    } else if (type == "audio") {
	src_pad = p;
    } else {
	return;
    }
    auto src_ghost = Gst::GhostPad::create(src_pad, "src");
    src_ghost->set_active();
            bin->add_pad(src_ghost);
}
*/

File: backend/src/branch.rs
use std::sync::{Arc,Mutex};
use gst::prelude::*;
use gst;
use std::str::Split;
use pad::SrcPad;
use signals::Signal;

struct CommonBranch {
    decoder: gst::Element,
    bin: gst::Bin,
    sink: gst::Pad,
}

impl CommonBranch {
    pub fn new () -> CommonBranch {
        let bin   = gst::Bin::new(None);
        let queue = gst::ElementFactory::make("queue", None).unwrap();

        for el in &["vaapih264dec", "vaapimpeg2dec", "vaapidecodebin"] {
            gst::ElementFactory::find(&el).map(|f| f.set_rank(0));
        };

        let decoder = gst::ElementFactory::make("decodebin",None).unwrap();

        queue.set_property("max-size-buffers", &20000);
        queue.set_property("max-size-bytes", &12000000);
        
        bin.add_many(&[&queue, &decoder]).unwrap();
        queue.link(&decoder).unwrap();

        let p = queue.get_static_pad("sink").unwrap();
        let sink_ghost = gst::GhostPad::new(Some("sink"), &p).unwrap();

        sink_ghost.set_active(true).unwrap();
        bin.add_pad(&sink_ghost).unwrap();

        CommonBranch { decoder, bin, sink: sink_ghost.upcast() }
    }
}

#[derive(Clone)]
pub struct VideoBranch {
    stream:   u32,
    channel:  u32,
    pid:      u32,
    pads:     Arc<Mutex<Vec<SrcPad>>>,
    analyser: gst::Element,
    decoder:  gst::Element,
    bin:      gst::Bin,
    sink:     gst::Pad,
    
    pub pad_added: Arc<Mutex<Signal<SrcPad>>>,
}

impl VideoBranch {
    pub fn new (stream: u32, channel: u32, pid: u32) -> VideoBranch {
        let common = CommonBranch::new();
        
        let pads      = Arc::new(Mutex::new(Vec::new()));
        let analyser  = gst::ElementFactory::make("videoanalysis",None).unwrap();
        let decoder   = common.decoder;
        let bin       = common.bin;
        let sink      = common.sink;
        let pad_added = Arc::new(Mutex::new(Signal::new()));

        let bin_c = bin.clone();
        let pads_c = pads.clone();
        let analyser_c = analyser.clone();
        let pad_added_c = pad_added.clone();
        
        decoder.connect_pad_added(move | _, pad | {
            let pcaps = String::from(pad.get_current_caps().unwrap().get_structure(0).unwrap().get_name());
            let caps_toks: Vec<&str> = pcaps.split('/').collect();
            if caps_toks.len() == 0 { return };

            let typ = &caps_toks[0];

            if *typ != "video" { return };

            let deint = gst::ElementFactory::make("deinterlace",None).unwrap();

            let sink_pad = deint.get_static_pad("sink").unwrap();
            let src_pad  = analyser_c.get_static_pad("src").unwrap();

            bin_c.add_many(&[&deint, &analyser_c]).unwrap();
            deint.link(&analyser_c).unwrap();

            deint.sync_state_with_parent();
            analyser_c.sync_state_with_parent();

            pad.link(&sink_pad);

            let spad = SrcPad::new(stream, channel, pid, "video", bin_c.clone(), src_pad);

            pad_added_c.lock().unwrap().emit(&spad);
            pads_c.lock().unwrap().push(spad);
        });
        
        VideoBranch { stream, channel, pid, pads, analyser, decoder, bin, sink, pad_added }
    }

    pub fn plug(&self, src_pad: &gst::Pad) {
        src_pad.link(&self.sink);
        self.bin.sync_state_with_parent();
    }

    pub fn add_to_pipe (&self, b: &gst::Bin) {
        b.add(&self.bin);
        self.bin.sync_state_with_parent();
    }
}

#[derive(Clone)]
pub struct AudioBranch {
    stream:   u32,
    channel:  u32,
    pid:      u32,
    pads:     Arc<Mutex<Vec<SrcPad>>>,
    analyser: gst::Element,
    decoder:  gst::Element,
    bin:      gst::Bin,
    sink:     gst::Pad,
    
    pub pad_added: Arc<Mutex<Signal<SrcPad>>>,
    pub audio_pad_added: Arc<Mutex<Signal<SrcPad>>>,
}

impl AudioBranch {
    pub fn new (stream: u32, channel: u32, pid: u32) -> AudioBranch {
        let common = CommonBranch::new();
        
        let pads      = Arc::new(Mutex::new(Vec::new()));
        let analyser  = gst::ElementFactory::make("audioanalysis",None).unwrap();
        let decoder   = common.decoder;
        let bin       = common.bin;
        let sink      = common.sink;
        let pad_added = Arc::new(Mutex::new(Signal::new()));
        let audio_pad_added = Arc::new(Mutex::new(Signal::new()));
        
        let bin_c = bin.clone();
        let pads_c = pads.clone();
        let analyser_c = analyser.clone();
        let pad_added_c = pad_added.clone();
        let audio_pad_added_c = audio_pad_added.clone();
        
        decoder.connect_pad_added(move | _, pad | {
            let pcaps = String::from(pad.get_current_caps().unwrap().get_structure(0).unwrap().get_name());
            let caps_toks: Vec<&str> = pcaps.split('/').collect();
            if caps_toks.len() == 0 { return };

            let typ = &caps_toks[0];

            if *typ != "audio" { return };

            let deint = gst::ElementFactory::make("deinterlace",None).unwrap();

            let sink_pad = deint.get_static_pad("sink").unwrap();
            let src_pad  = analyser_c.get_static_pad("src").unwrap();

            bin_c.add_many(&[&deint, &analyser_c]).unwrap();
            deint.link(&analyser_c).unwrap();

            deint.sync_state_with_parent();
            analyser_c.sync_state_with_parent();

            pad.link(&sink_pad);

            let spad = SrcPad::new(stream, channel, pid, "audio", bin_c.clone(), src_pad);

            pad_added_c.lock().unwrap().emit(&spad);
            audio_pad_added_c.lock().unwrap().emit(&spad);
            pads_c.lock().unwrap().push(spad);
        });

        bin.sync_children_states();
        AudioBranch { stream, channel, pid, pads, analyser, decoder, bin, sink, pad_added, audio_pad_added }
    }

    pub fn plug(&self, src_pad: &gst::Pad) {
        src_pad.link(&self.sink);
        self.bin.sync_state_with_parent();
    }

    pub fn add_to_pipe (&self, b: &gst::Bin) {
        b.add(&self.bin);
        self.bin.sync_state_with_parent();
    }
}

#[derive(Clone)]
pub enum Branch {
    Video(VideoBranch),
    Audio(AudioBranch),
}

impl Branch {

    pub fn new(stream: u32, channel: u32, pid: u32, typ: &str) -> Option<Branch> {
        match typ {
            "video" => Some(Branch::Video(VideoBranch::new(stream, channel, pid))),
           // "audio" => Some(Branch::Audio(AudioBranch::new(stream, channel, pid))),
            _       => None
        }
    }
    
    pub fn plug(&self, src_pad: &gst::Pad) {
        match *self {
            Branch::Video(ref br) => br.plug(src_pad),
            Branch::Audio(ref br) => br.plug(src_pad),
        };
    }

    pub fn add_to_pipe (&self, b: &gst::Bin) {
        match *self {
            Branch::Video(ref br) => br.add_to_pipe(b),
            Branch::Audio(ref br) => br.add_to_pipe(b),
        }
    }    
}